### ISSUE TITLE
Duplicate method name after compilation

### ISSUE THREAD

**Issue by kubukoz** [lampepfl/dotty#19274]
## Compiler version

3.3.1

## Minimized code

```scala
//> using scala 3.3.1

case class StaticBinding(v: String) {
  private def copy$default$1(): String = ???
}

object Demo extends App {
  val b = StaticBinding("test")
  println(b)
}

```

## Output

```scala
Error: Unable to initialize main class Demo
Caused by: java.lang.ClassFormatError: Duplicate method name "copy$default$1" with signature "()Ljava.lang.String;" in class file StaticBinding
Program exited with return code 1.
```

## Expectation

Compilation failure (same as in 2.13.11):

```
[error] ./main.scala:3:26
[error] double definition:
[error] private def copy$default$1(): String at line 4 and
[error] def copy$default$1: String at line 3
[error] have same type
[error] case class StaticBinding(v: String) {
[error]                          ^
Error compiling project (Scala 2.13.11, JVM (17))
```

**Comment by som-snytt** [comment]
Noticed at erasure in Scala 2. The error makes me appreciate nicely formatted messages, as well as informative ones.

**Comment by soronpo** [comment]
User code is not supposed to have names with `$` in them.

**Comment by kubukoz** [comment]
> User code is not supposed to have names with `$` in them.

Then it should at least be a warning ;) is this documented somewhere? I've seen [a similar comment before](https://github.com/lampepfl/dotty/issues/12469), but it seems like an unwritten (or at least not loud and enforced) rule.

For context, I was trying to do this to keep binary compatibility after a parameter type change in a case class. Isn't this a legitimate thing to do as a library author trying to avoid breaking changes?

**Comment by som-snytt** [comment]
@kubukoz yes, no dollars is at the beginning of the spec, not obscure at all. That's not only to avoid collisions, but I assume also to avoid anyone relying on an implementation detail.

Your use case is interesting, like "can I break the law if it's to keep someone from breaking the law?" There is even something selfless and noble about it.

**Comment by som-snytt** [comment]
The ticket for making this harder to fall into is https://github.com/scala/scala3/issues/18234.

The reason for the behavior is that your `copy$default$1` is a `SimpleName` and the real one is derived (structured), so the methods don't compare as matching.

**Comment by SolalPirelli** [comment]
Closing since we now warn about this.
